### PR TITLE
Refresh token on credential rotation

### PR DESCRIPTION
**Pull request by howardjohn**
Currently we close the SDS connection on rotation. For short lived
certs, we probably shouldn't do this. Instead, we can just refresh the
token in the standard case where we just need to read the jwt file
again.

We are explicitly reading the file and not just storing the content
somewhere because the jwt token will refresh as well.

### CODE REVIEW

**Comment by howardjohn** [comment]
this seems to work but I didn't add any tests yet

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -553,9 +557,18 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
 			// Send the notification to close the stream if token is expired, so that client could re-connect with a new token.
 			if sc.isTokenExpired() {
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/cache/secretcache.go]: The shouldRefresh() and isTokenExpired() is a bit confusing to me. Perhaps we need to consolidate them in the future..

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -553,9 +557,18 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
...
 				cacheLog.Debugf("%s token expired", logPrefix)
-				sc.callbackWithTimeout(connKey, nil /*nil indicates close the streaming connection to proxy*/)
-
-				return true
+				if fetchToken != nil {
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/cache/secretcache.go]: Is it okay for you to add an test with short lived token, and verify that the token refresh logic works as expected?

**Comment by howardjohn** [comment]
@myidpt I guess this missed 1.5, but we should get this working in master I think. Let me know if you have any thoughts

**Comment by howardjohn** [comment]
ping @istio/wg-security-maintainers

**Comment by howardjohn** [comment]
ping @istio/wg-security-maintainers

**Comment by howardjohn** [comment]
ping @istio/wg-security-maintainers

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -553,9 +557,18 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
...
+					t, err := fetchToken()
+					if err != nil {
+						cacheLog.Errorf("failed to fetch new token, closing connection: %v", err)
+						sc.callbackWithTimeout(connKey, nil /*nil indicates close the streaming connection to proxy*/)
+						return true
+					}
+					e.Token = t
```
Comment (myidpt) [inline security/pkg/nodeagent/cache/secretcache.go]: I think you still need to call sc.callbackWithTimeout() with the updated secret, right?

**Comment by myidpt** [comment]
Sorry for the late reply. Thanks for the work. Left one comment.

**Comment by howardjohn** [comment]
Abandoning this PR, I had time to work on this 1 month ago when I submitted it but I no longer do

**Pull request closed without merging**